packages/react/src/components/TextInput/TextInput.js

```diff
@@ -63,33 +95,83 @@ const TextInput = React.forwardRef(function TextInput(
   const helperTextClasses = classNames(`${prefix}--form__helper-text`, {
     [`${prefix}--form__helper-text--disabled`]: other.disabled,
   });
-  const label = labelText ? (
-    <label htmlFor={id} className={labelClasses}>
-      {labelText}
-    </label>
-  ) : null;
+  const TextInputLabel = () => {
+    const labelContent = labelText && (
+      <label htmlFor={id} className={labelClasses}>
+        {labelText}
+      </label>
+    );
+    if (labelContent && useCharCount) {
+      return (
+        <div className={`${prefix}--text-input__character-counter-title`}>
+          {labelContent}
+          <CharCounter
+            disabled={other.disabled}
+            count={value.length}
+            maxLength={maxLength}
+          />
+        </div>
+      );
+    }
+    return labelContent;
+  };
   const error = invalid ? (
     <div className={`${prefix}--form-requirement`} id={errorId}>
       {invalidText}
     </div>
   ) : null;
-  const input = (
-    <input {...textInputProps({ invalid, sharedTextInputProps, errorId })} />
+  const inputField = (
+    <input
+      {...textInputProps({ invalid, sharedTextInputProps, errorId })}
+      value={value}
+      onInput={e => setInput(e.target.value)}
+    />
   );
-  const helper = helperText ? (
-    <div className={helperTextClasses}>{helperText}</div>
-  ) : null;
+  const input =
+    readOnly && value ? (
+      <Tooltip showIcon={false} triggerText={inputField}>
```

Comment:
Is the usage of interactive tooltip here from the UI spec our designer gave us?

Reply:
yeah the tooltip is for showing the full value of the input field when it is in read only mode (ref https://github.com/carbon-design-system/carbon/issues/2177)

Reply:
Thanks, looking at https://user-images.githubusercontent.com/15144993/57665202-9970fc00-75c0-11e9-9a3f-b628d7b168a9.png - Probably my specific question is to double-check if it's an interactive tooltip, instead of the hover-over variant (definition/icon)?

Reply:
the reason for interactive tooltip is because the content could be several lines long, which isn't suggested for icon or definition tooltips IIRC. but we may need to make some updates due to #3812